### PR TITLE
UI redesign: Share menu

### PR DESCRIPTION
**Pull request by kandre**
This PR updates our Share menu by using [Clipboard API](https://w3c.github.io/clipboard-apis/) with [Html5rocks solution](http://updates.html5rocks.com/2015/04/cut-and-copy-commands).

Since Document.execCommand() is newly supported by latest browsers version, this solution works only from:

- IE10+
- Chrome 43+

However, for now:
- Firefox supports this but requires a preference change [See cut and copy commands here](Firefox supports this but requires a preference change). Without this Firefox throws an error.
- For now, it does not work on Safari.

For non-compatible browsers, we could eventually inform the user that the copy to clipboard should be done manually.

### CODE REVIEW

**Comment by cedricmoullet** [comment]
Test link ?

**Comment by cedricmoullet** [comment]
We should provide a feedback somewhere: "The link has been copied to the clipboard"

**Comment by kandre** [comment]
jenkins test

**Comment by ponceta** [comment]
jenkins down

**Comment by kandre** [comment]
jenkins test

**Comment by kandre** [comment]
jenkins test

**Comment by kandre** [comment]
I removed some parts of the test related to the permalink, because with this PR, the permalink we always provide is a shortened one (hence no information of coordinates to test with browserstack). Besides, we could eventually define a permalink test (issue #2347)

**Comment by gjn** [comment]
The permalink url is visible in other parts of the page as well (for instance on the mobile/desktop switcher)...just loook for it.

The tests should completely work inside this PR.

File: test/selenium/search_test.js [hunk 1]
```diff
@@ -21,12 +19,6 @@ var runTest = function(cap, driver, target){
         assert.ok(url.indexOf(QUERYSTRING_OF_BERN) > -1);
     });
   }
-  //And have a look at the permanent Link
-  driver.findElement(webdriver.By.xpath("//*[@ng-model='permalinkValue']")).getAttribute("value").then(function(val){
-      //The perma Link should point to Bern
-      assert.ok(val.indexOf(QUERYSTRING_OF_BERN) > -1);
-  });
-  //We click on the "share" button what closes the share menu
```
Comment (gjn) [inline test/selenium/search_test.js]: Please don't remove this, as without this, the test is not complete.

File: test/selenium/swisssearch_test.js [hunk 2]
```diff
@@ -14,10 +14,6 @@ var runTest = function(cap, driver, target){
   driver.findElement(webdriver.By.xpath("//*[contains(text(), 'Raron')]"));
   driver.findElement(webdriver.By.xpath("//*[contains(text(), 'Flugfeld')]")).click();
   driver.findElement(webdriver.By.xpath("//a[contains(@href, '" + QUERYSTRING_OF_RARON + "')]"));
-  //parameter should disappear when selection is done
-  driver.findElement(webdriver.By.xpath("//input[@ng-model='permalinkValue']")).getAttribute("value").then(function(val) {
-    assert.ok(val.indexOf('swisssearch') == -1);
-  });
```
Comment (gjn) [inline test/selenium/swisssearch_test.js]: idem

File: test/selenium/swisssearch_test.js [hunk 3]
```diff
@@ -26,10 +22,6 @@ var runTest = function(cap, driver, target){
   driver.findElement(webdriver.By.xpath("//*[contains(text(), 'Standorte')]"));
   driver.findElement(webdriver.By.xpath("//*[contains(text(), 'Karten')]"));
   driver.findElement(webdriver.By.xpath("//i[@ng-click='clearInput()']")).click();
-  //parameter should disappear when selection is done
-  driver.findElement(webdriver.By.xpath("//input[@ng-model='permalinkValue']")).getAttribute("value").then(function(val) {
-    assert.ok(val.indexOf('swisssearch') == -1);
-  });
```
Comment (gjn) [inline test/selenium/swisssearch_test.js]: idem

File: test/selenium/swisssearch_test.js [hunk 4]
```diff
@@ -37,18 +29,9 @@ var runTest = function(cap, driver, target){
     //wait until topics related stuff is loaded. We know this when catalog is there
     driver.findElement(webdriver.By.xpath("//a[contains(text(), 'Grundlagen und Planung')]"));
     driver.findElement(webdriver.By.xpath("//a[contains(@href, '" + QUERYSTRING_MOOS + "')]"));
-    //parameter stays after initial automatic selection
-    driver.findElement(webdriver.By.xpath("//input[@ng-model='permalinkValue']")).getAttribute("value").then(function(val) {
-      assert.ok(val.indexOf(QUERYSTRING_MOOS) > -1);
-      assert.ok(val.indexOf('swisssearch=br') > -1);
-    });
```
Comment (gjn) [inline test/selenium/swisssearch_test.js]: idem

File: test/selenium/swisssearch_test.js [hunk 5]
```diff
@@ -37,18 +29,9 @@ var runTest = function(cap, driver, target){
...
     //parameter is removed by map action (simulating zoom here)
     driver.findElement(webdriver.By.xpath("//button[@class='ol-zoom-in']")).click();
     driver.findElement(webdriver.By.xpath("//a[contains(@href, 'zoom=11')]"));
-    //parameter should disappear when selection is done
-    driver.findElement(webdriver.By.xpath("//input[@ng-model='permalinkValue']")).getAttribute("value").then(function(val) {
-      assert.ok(val.indexOf('swisssearch') == -1);
-    });
```
Comment (gjn) [inline test/selenium/swisssearch_test.js]: idem

**Comment by kandre** [comment]
Good point, thks for the info

**Comment by cedricmoullet** [comment]
@kandre Could you please provide a test link ?

**Comment by kandre** [comment]
Here is a [test link](https://mf-geoadmin3.dev.bgdi.ch/kan_UI_share_menu).

**Comment by kandre** [comment]
I added the by default hidden pane for embed link in 'More options'. As soon as we are satisfied with such solution, I will ask for better translations

**Comment by gjn** [comment]
@kandre Status of this? Rebase and squash?

**Comment by kandre** [comment]
Some stuff to be fixed still. I ll remove the WIP when ok

**Comment by gjn** [comment]
squash?

**Comment by kandre** [comment]
Hard to solve a bug on Apple mobiles: To autoselect text in input, I use setSelectionRange which is fully cross-compatible (tested, I confirm). However cant make it work on safari mobiles (only) within a success function of a $http. 

@gjn have you already encounter such issue ?

**Comment by kandre** [comment]
My problem seems to come from a safari (mobile) feature, different from other browsers.

Actually, according to [this](http://stackoverflow.com/questions/6287478/mobile-safari-autofocus-text-field/7332160#7332160), 'iOS will only allow focus to be triggered on other elements, from within a function, if the first function in the call stack was triggered by a non-programmatic event'

I think I have no options than doing doing a hack or not autoselect input on safari mobile.

**Comment by kandre** [comment]
Autoselect works on mobiles except on safari mobiles. However it is not convenient because browser does not understand that it is a user event interaction, and hence does not propose to copy/paste it (native copy/paste). Hence, no autoselect on mobiles.

**Comment by kandre** [comment]
jenkins test

**Comment by kandre** [comment]
Squashed and ok for review for the code side. Some translation are still missing.

**Comment by kandre** [comment]
jenkins test

File: src/components/BrowserSnifferService.js [hunk 6]
```diff
@@ -28,6 +28,12 @@
       var webkit = !msie && /WebKit/.test(ua);
       var opera = !msie && /(OPiOS|OPR)\//.test(ua);
       var chrome = !msie && !opera && /(CriOS|Chrome)\//.test(ua);
+      if (chrome) {
+        var m = ua.match(/Chrome\/(\d+)\./);
```
Comment (oterral) [inline src/components/BrowserSnifferService.js]: What's happening when useragent contains `CriOS` ?

File: src/components/BrowserSnifferService.js [hunk 7]
```diff
@@ -103,6 +109,7 @@
         webkit: webkit,
         mac: mac,
         safari: safari,
+        chrome: chrome,
```
Comment (oterral) [inline src/components/BrowserSnifferService.js]: Put a comment  like for the ios property : 
 `// false or chrome version number

File: src/components/share/ShareDirective.js [hunk 8]
```diff
@@ -10,10 +10,12 @@
   ]);
 
   module.directive('gaShare',
-      function($http, $window, gaPermalink, gaBrowserSniffer) {
+      function($document, $http, $translate, $window, gaPermalink,
+          gaBrowserSniffer) {
           return {
```
Comment (oterral) [inline src/components/share/ShareDirective.js]: weird indentation should be 2

File: src/components/share/ShareDirective.js [hunk 9]
```diff
@@ -21,15 +23,25 @@
...
+
+                $('#permalinkInput').tooltip({
+                  placement: 'bottom'
+                });
               }
+
               $('.ga-share-permalink input').on({
                 focus: function() {
-                  this.setSelectionRange(0, 9999);
+                  if (!gaBrowserSniffer.mobile) {
```
Comment (oterral) [inline src/components/share/ShareDirective.js]: No focus on mobile ?  in that case you can put this event in the previous `if  (!gaBrowserSniffer.mobile) `

File: src/components/share/ShareDirective.js [hunk 10]
```diff
@@ -69,13 +81,47 @@
...
+
+                // Remove the selections - NOTE: the two following lines
+                // do the job for all three browsers
+                $document[0].getSelection().removeAllRanges();
+                $document[0].getSelection().addRange($document[0]
+                    .createRange());
+              };
+
               // Select the input field on click in order to allow copy/paste
               scope.selectOnClick = function(e) {
```
Comment (oterral) [inline src/components/share/ShareDirective.js]: The click event  does the same thing  of the focus event  on l41 ? In which the click event is called and not the focus event ?

File: src/components/share/partials/share.html [hunk 11]
```diff
@@ -1,15 +1,17 @@
 <div class="ga-share-icons">
   <a class="ga-share-icon" target="_blank"
+    data-original-title="{{'mail_tooltip' | translate}}"
+    ng-href="mailto:?body={{encodedPermalinkHref}}">
```
Comment (oterral) [inline src/components/share/partials/share.html]: Is an `href` attribute enough ?

File: src/locales/de.json [hunk 12]
```diff
@@ -39,6 +39,9 @@
 	"connect": "Verbinden",
 	"contact_us": "Kontakt",
 	"coordinates_label": "Koordinaten (m)",
+  "copy_url_fail": "Link has not been copied!",
```
Comment (oterral) [inline src/locales/de.json]: Weird indent in all translations files

File: src/components/share/ShareDirective.js [hunk 13]
```diff
@@ -21,15 +23,25 @@
               var shortenUrl = scope.options.shortenUrl;
               scope.qrcodegeneratorPath = scope.options.qrcodegeneratorPath;
               scope.mobile = gaBrowserSniffer.mobile;
+              scope.gteIE11 = (gaBrowserSniffer.msie >= 11);
+              scope.gteChrome42 = (gaBrowserSniffer.chrome >= 42);
```
Comment (oterral) [inline src/components/share/ShareDirective.js]: Only one scope variable is enough  : 
`iscope.isCopyAllow = (gaBrowserSniffer.msie >= 11) &&(gaBrowserSniffer.chrome >= 42) `

File: src/components/share/partials/share.html [hunk 14]
```diff
@@ -18,45 +20,56 @@
...
+      <button class="btn btn-default" ng-click="copyUrl('permalink');" translate>copy_url</button>
     </span>
   </div>
-  <div translate>share_disclaimer</div>  
 </div>
-      
+
+<div class="ga-message">{{userMessage}}</div>
+
+<div class="show-more" ng-if="!mobile" ng-click="showMore=!showMore">
```
Comment (oterral) [inline src/components/share/partials/share.html]: Add a `ga` prefix pls : `ga-show-more`

**Comment by oterral** [comment]
@kandre there is som missing css int the iframe modal popup, iframe input and select box are not on the same line.

**Comment by oterral** [comment]
2 other things:

- when we click on "More options ... " the parent element should be scrolled  to see the new options.
- in the iframe modal popup the " Copy link" button appears on non supported browser

File: src/components/share/ShareDirective.js [hunk 15]
```diff
@@ -69,13 +81,47 @@
                 }).success(function(response) {
                   scope.permalinkValue = response.shorturl;
                   scope.urlShortened = true;
+                  scope.$applyAsync(function() {
+                    // Auto-select the shortened permalink (not on mobiles)
+                    if (!gaBrowserSniffer.mobile) {
```
Comment (oterral) [inline src/components/share/ShareDirective.js]: If not mobile no need to call `$applyAsync`

File: src/components/share/ShareDirective.js [hunk 16]
```diff
@@ -69,13 +81,47 @@
...
               };
 
+              // Use clipboard API to copy URL in OS clipboard
+              scope.copyUrl = function(type) {
+                // Select the permalink anchor text
+                if (type == 'permalink') {
+                  var inputElement = $('#permalinkInput');
+                } else if (type == 'embed') {
+                   var inputElement = $('.ga-embed-input');
+                }
```
Comment (oterral) [inline src/components/share/ShareDirective.js]: this `if` doesn't look very generic to me. Why not create a `ga-copy`  directive which takes a  jquery selector as parameter : 

`<button ga-copy=".ga-embed-input"></button>`

File: src/components/BrowserSnifferService.js (hunk 6 above)
Comment (kandre) [inline src/components/BrowserSnifferService.js]: chrome set as null...I correct it

Comment (oterral) [inline src/components/BrowserSnifferService.js]: Can I suggest to write tests for this ;)

Comment (kandre) [inline src/components/BrowserSnifferService.js]: you mean adapt https://github.com/geoadmin/mf-geoadmin3/blob/master/test/specs/BrowserSnifferService.spec.js
accordingly ?

File: src/components/share/ShareDirective.js (hunk 15 above)
Comment (kandre) [inline src/components/share/ShareDirective.js]: just here, this applyAsync is still needed in order to keep autoselect applied after ajax success

Comment (oterral) [inline src/components/share/ShareDirective.js]: yes you're right

**Comment by kandre** [comment]
I changed the cursor style when on input element

File: src/components/share/ShareDirective.js (hunk 16 above)
Comment (oterral) [inline src/components/share/ShareDirective.js]: Forget this we'll see in the future if it's necessary

**Comment by kandre** [comment]
Here is a [test link](https://mf-geoadmin3.dev.bgdi.ch/kan_UI_share_menu).

**Comment by cedricmoullet** [comment]
copy_url is not translated

2015-06-02 18:39 GMT+02:00 kandre <notifications@github.com>:

> Here is a test link <https://mf-geoadmin3.dev.bgdi.ch/kan_UI_share_menu>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geoadmin/mf-geoadmin3/pull/2314#issuecomment-108009260>
> .
>



-- 
Twitter: http://twitter.com/cedricmoullet
Linked In: http://www.linkedin.com/in/cedricmoullet

<http://twitter.com/cedricmoullet>

**Comment by cedricmoullet** [comment]
UI issue on Mac / Chrome: 
image

**Comment by cedricmoullet** [comment]
copy_success is also not translated

**Comment by kandre** [comment]
I redeployed the branch. It works now

**Comment by gjn** [comment]
Could you please squash it. Then we'll merge.

**Comment by kandre** [comment]
Squashed !

File: src/components/BrowserSnifferService.js [hunk 17]
```diff
@@ -28,6 +28,12 @@
...
       var opera = !msie && /(OPiOS|OPR)\//.test(ua);
       var chrome = !msie && !opera && /(CriOS|Chrome)\//.test(ua);
+      if (chrome) {
+        var m = ua.match(/(CriOS|Chrome)\/(\d+)\./);
+        if (m && m.length > 2) {
+          chrome = parseInt(m[2], 10);
+        }
+      }
       var safari = !msie && !opera && !chrome && /Safari/.test(ua);
       var iosChrome = ios && chrome;
```
Comment (oterral) [inline src/components/BrowserSnifferService.js]: Move it before the `íf (chrome)`

File: src/locales/rm.json [hunk 18]
```diff
@@ -242,12 +244,15 @@
 	"send": "Trametter",
 	"settings": "Configuraziun ",
 	"share": "Parter",
-	"share_disclaimer": "Durch das Einbetten dieser Karte stimmen Sie den  <a href='http://www.geo.admin.ch/internet/geoportal/rm/home/geoadmin/contact.html#iFrame' target='_blank'>Nutzungsbedingungen</a> zu.",
+	"share_disclaimer": "Sie können die Karte in Ihre Website oder ein Blog einbetten. <a href='http://www.geo.admin.ch/internet/geoportal/rm/home/geoadmin/contact.html#iFrame' target='_blank'>Nutzungsbedingungen</a>",
 	"share_file_disclaimer": "Ihre Zeichnung wird  automatisch für ein Jahr gespeichert. Durch  die Nutzung  dieses Dienstes stimmen Sie den  <a href='http://www.geo.admin.ch/internet/geoportal/de/home/geoadmin/contact.html#drawing ' target='_blank'>Nutzungsbedingungen</a> zu.",
-	"share_file_link_descr_admin": "",
+	"share_file_link_descr_admin": "share_file_link_descr_admin",
```
Comment (oterral) [inline src/locales/rm.json]: Remove it if unused

File: src/components/BrowserSnifferService.js (hunk 17 above)
Comment (kandre) [inline src/components/BrowserSnifferService.js]: Yes no version number for iosChrome

File: src/locales/rm.json (hunk 18 above)
Comment (kandre) [inline src/locales/rm.json]: I dont know. Doesn't it come from another PR (kml administration) ?

**Comment by kandre** [comment]
Thanks a lot for reviews

File: src/locales/rm.json (hunk 18 above)
Comment (oterral) [inline src/locales/rm.json]: yes something is weird but it's another PR

**Comment by oterral** [comment]
thx for the hard work

**Pull request merged**